swap/simulations_test.go

```diff
@@ -0,0 +1,779 @@
+						} else {
+							tp.Send(ctx, &testMsgBySender{})
+						}
+						msgCount++
+					} else {
+						break ITER
+					}
+				}
+			}
+		}
+
+		// before we can check the balances, we need to wait a bit, as the last messages
+		// may still be processed
+		select {
+		case <-ctx.Done():
+			return errors.New("timed out waiting for all messages to arrive, aborting")
+		case <-allMessagesArrived:
+		}
+		log.Debug("all messages arrived")
+
+		//now iterate again and check that every node has the same
+		//balance with a peer as that peer with the same node,
+		//but in inverted signs
+		for _, node := range nodes {
+			item, ok := sim.NodeItem(node, bucketKeySwap)
+			if !ok {
+				return errors.New("no swap in simulation bucket")
+			}
+			ts := item.(*testService)
+			// for each node look up the peers
+			for _, p := range nodes {
+				// no need to check self
+				if p == node {
+					continue
+				}
+
+				peerItem, ok := sim.NodeItem(p, bucketKeySwap)
+				if !ok {
+					return errors.New("no swap in simulation bucket")
+				}
+				peerTs := peerItem.(*testService)
+
+				// balance of the node with peer p
+				nodeBalanceWithP, err := ts.swap.loadBalance(p)
+				if err != nil {
+					return fmt.Errorf("expected balance for peer %v to be found, but not found", p)
+				}
+				// balance of the peer with node
+				pBalanceWithNode, err := peerTs.swap.loadBalance(node)
+				if err != nil {
+					return fmt.Errorf("expected counter balance for node %v to be found, but not found", node)
+				}
+				if nodeBalanceWithP != -pBalanceWithNode {
+					return fmt.Errorf("Expected symmetric balances, but they are not: %d vs %d", nodeBalanceWithP, pBalanceWithNode)
+				}
+			}
+		}
+
+		return nil
+
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	counter.Clear()
+	log.Info("Simulation ended")
+}
+
+func waitForChequeProcessed(ts *testService) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
+	defer cancel()
+
+	backend := ts.swap.backend.(*swapTestBackend)
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("timed out waiting for cheque to be processed")
+	case <-backend.cashDone:
+		return nil
+	}
+}
+
+func (ts *testService) Protocols() []p2p.Protocol {
+	spec := newTestSpec()
+	return []p2p.Protocol{
+		{
+			Name:    spec.Name,
+			Version: spec.Version,
+			Length:  spec.Length(),
+			Run:     ts.runProtocol,
+		},
+		{
+			Name:    Spec.Name,
+			Version: Spec.Version,
+			Length:  Spec.Length(),
+			Run:     ts.swap.run,
+		},
+	}
+}
+
+// APIs retrieves the list of RPC descriptors the service provides
+func (ts *testService) APIs() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "testAccounted",
+			Version:   "1.0",
+			Service:   ts,
+			Public:    false,
+		},
+	}
+}
+
+// runProtocol for the test spec
+func (ts *testService) runProtocol(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	peer := protocols.NewPeer(p, rw, ts.spec)
+	tp := &testPeer{Peer: peer}
+	ts.peers[tp.ID()] = tp
```

Comment:
This still leads to a `concurrent map writes` error sometimes.

Reply:
Adressed with locks